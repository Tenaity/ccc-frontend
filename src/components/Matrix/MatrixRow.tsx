import React from "react";
import Badge from "../Badge";
import type { ShiftCode, Staff, Position } from "../../types";
import { fmtYMD, getDow, isWeekend } from "../../utils/date";

const tdCenter = { borderTop: "1px solid #f0f0f0", padding: "6px 8px", textAlign: "center" as const, whiteSpace: "nowrap" as const };
const tdWeekend = { background: "#FFFDF0" };
const tdStickyLeft = { ...tdCenter, position: "sticky" as const, left: 0, zIndex: 1, background: "#fff", textAlign: "left" as const };
type Cell = { code: ShiftCode; position: Position | null };

export default function MatrixRow({
    staff, index, year, month, days,
    assignmentIndex, summariesByStaffId
}: {
    staff: Staff; index: number; year: number; month: number; days: number[];
    assignmentIndex: Map<string, Cell>;
    summariesByStaffId: Map<number, { counts: Record<string, number>, credit: number, dayCount: number, nightCount: number }>;
}) {
    const sum = summariesByStaffId.get(staff.id) || { counts: {}, credit: 0, dayCount: 0, nightCount: 0 };

    return (
        <tr style={{ background: index % 2 ? "#fff" : "#fbfbfd" }}>
            <td className="sticky-col" style={tdStickyLeft}>
                <div style={{ fontWeight: 600 }}>{staff.full_name}</div>
                <div style={{ fontSize: 12, color: "#666" }}>{staff.role} {staff.can_night ? "" : "· không đêm"}</div>
            </td>

            {days.map(d => {
                const dateKey = fmtYMD(year, month, d);
                const cell = assignmentIndex.get(`${staff.id}|${dateKey}`);
                const code = cell?.code ?? "";
                const pos = cell?.position ?? null;
                const wk = isWeekend(getDow(year, month, d));

                const isLeaderDay = code === "K" && pos === "TD";
                const isLeaderNight = code === "Đ" && pos === "TD" && staff.role === "TC";

                return (
                    <td key={d} style={{ ...tdCenter, ...(wk ? tdWeekend : null) }}>
                        <Badge
                            code={code}
                            crown={isLeaderDay || isLeaderNight}
                            variant={
                                pos === "K_WHITE" ? "k-white" :
                                    pos === "PGD" ? "pgd" :
                                        isLeaderDay ? "leader-day" :
                                            isLeaderNight ? "leader-night" : "td"
                            }
                        />
                    </td>
                );
            })}

            <td style={tdCenter}>{sum.counts?.CA1 || 0}</td>
            <td style={tdCenter}>{sum.counts?.CA2 || 0}</td>
            <td style={tdCenter}>{sum.counts?.K || 0}</td>
            <td style={tdCenter}>{sum.counts?.HC || 0}</td>
            <td style={tdCenter}>{sum.counts?.Đ || 0}</td>
            <td style={tdCenter}>{sum.counts?.P || 0}</td>
            <td style={tdCenter}>{sum.dayCount}</td>
            <td style={tdCenter}>{sum.nightCount}</td>
            <td style={{ ...tdCenter, fontWeight: 700, background: "#f9fafb" }}>{sum.credit}</td>
        </tr>
    );
}
