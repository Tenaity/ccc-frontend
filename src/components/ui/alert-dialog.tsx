import * as AlertDialogPrimitive from "@radix-ui/react-alert-dialog"
import * as React from "react"

import { cn } from "@/lib/utils"
import { buttonVariants } from "@/components/ui/button"

const AlertDialog = AlertDialogPrimitive.Root

const AlertDialogTrigger = AlertDialogPrimitive.Trigger

const AlertDialogPortal = AlertDialogPrimitive.Portal

const AlertDialogOverlay = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Overlay>,
  React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Overlay>
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Overlay
    ref={ref}
    className={cn(
      "fixed inset-0 z-50 bg-black/50 backdrop-blur-sm data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=open]:fade-in-0 data-[state=closed]:fade-out-0",
      className
    )}
    {...props}
  />
))
AlertDialogOverlay.displayName = AlertDialogPrimitive.Overlay.displayName

const AlertDialogContent = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Content>,
  React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Content>
>(({ className, children, ...props }, ref) => (
  <AlertDialogPortal>
    <AlertDialogOverlay />
    <AlertDialogPrimitive.Content
      ref={ref}
      className={cn(
        "fixed left-1/2 top-1/2 z-50 w-full max-w-md -translate-x-1/2 -translate-y-1/2 rounded-2xl border border-border/60 bg-background p-6 shadow-2xl outline-none",
        "data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=open]:fade-in-0 data-[state=open]:zoom-in-95 data-[state=closed]:fade-out-0 data-[state=closed]:zoom-out-95",
        className
      )}
      {...props}
    >
      {children}
    </AlertDialogPrimitive.Content>
  </AlertDialogPortal>
))
AlertDialogContent.displayName = AlertDialogPrimitive.Content.displayName

const AlertDialogHeader = ({
  className,
  ...props
}: React.HTMLAttributes<HTMLDivElement>) => (
  <div className={cn("flex flex-col gap-2 text-center sm:text-left", className)} {...props} />
)
AlertDialogHeader.displayName = "AlertDialogHeader"

const AlertDialogFooter = ({
  className,
  ...props
}: React.HTMLAttributes<HTMLDivElement>) => (
  <div
    className={cn(
      "flex flex-col-reverse gap-2 pt-4 sm:flex-row sm:justify-end",
      className
    )}
    {...props}
  />
)
AlertDialogFooter.displayName = "AlertDialogFooter"

const AlertDialogTitle = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Title>,
  React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Title>
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Title
    ref={ref}
    className={cn("text-lg font-semibold leading-none tracking-tight", className)}
    {...props}
  />
))
AlertDialogTitle.displayName = AlertDialogPrimitive.Title.displayName

const AlertDialogDescription = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Description>,
  React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Description>
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Description
    ref={ref}
    className={cn("text-sm text-muted-foreground", className)}
    {...props}
  />
))
AlertDialogDescription.displayName = AlertDialogPrimitive.Description.displayName

const AlertDialogAction = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Action>,
  React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Action>
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Action
    ref={ref}
    className={cn(buttonVariants({ variant: "default" }), "sm:ml-3", className)}
    {...props}
  />
))
AlertDialogAction.displayName = AlertDialogPrimitive.Action.displayName

const AlertDialogCancel = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Cancel>,
  React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Cancel>
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Cancel
    ref={ref}
    className={cn(buttonVariants({ variant: "outline" }), className)}
    {...props}
  />
))
AlertDialogCancel.displayName = AlertDialogPrimitive.Cancel.displayName

export {
  AlertDialog,
  AlertDialogTrigger,
  AlertDialogPortal,
  AlertDialogOverlay,
  AlertDialogContent,
  AlertDialogHeader,
  AlertDialogFooter,
  AlertDialogTitle,
  AlertDialogDescription,
  AlertDialogAction,
  AlertDialogCancel,
}
