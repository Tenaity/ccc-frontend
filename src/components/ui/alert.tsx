import { cva, type VariantProps } from "class-variance-authority"
import * as React from "react"

import { cn } from "@/lib/utils"

const alertVariants = cva(
  "relative w-full rounded-lg border px-4 py-3 text-sm [&>svg]:absolute [&>svg]:left-4 [&>svg]:top-4 [&>svg]:text-foreground [&>svg+div]:translate-y-[-2px] [&>svg~*]:pl-8",
  {
    variants: {
      variant: {
        default: "bg-background text-foreground",
        success: "border-emerald-200 bg-emerald-50 text-emerald-900 dark:border-emerald-900/40 dark:bg-emerald-950 dark:text-emerald-50",
        destructive:
          "border-destructive/60 bg-destructive/10 text-destructive dark:text-destructive",
        warning:
          "border-amber-200 bg-amber-50 text-amber-900 dark:border-amber-900/60 dark:bg-amber-950 dark:text-amber-50",
        muted: "border-border/60 bg-muted text-muted-foreground",
      },
    },
    defaultVariants: {
      variant: "default",
    },
  }
)

export interface AlertProps
  extends React.ComponentPropsWithoutRef<"div">,
    VariantProps<typeof alertVariants> {}

const Alert = React.forwardRef<HTMLDivElement, AlertProps>(
  ({ className, variant, children, ...props }, ref) => (
    <div
      ref={ref}
      role="alert"
      className={cn(alertVariants({ variant }), className)}
      {...props}
    >
      {children}
    </div>
  )
)
Alert.displayName = "Alert"

const AlertTitle = React.forwardRef<
  React.ElementRef<"h5">,
  React.ComponentPropsWithoutRef<"h5">
>(({ className, children, ...props }, ref) => (
  <h5
    ref={ref}
    className={cn("mb-1 font-semibold leading-none tracking-tight", className)}
    {...props}
  >
    {children}
  </h5>
))
AlertTitle.displayName = "AlertTitle"

const AlertDescription = React.forwardRef<
  HTMLParagraphElement,
  React.ComponentPropsWithoutRef<"p">
>(({ className, children, ...props }, ref) => (
  <p
    ref={ref}
    className={cn("text-sm leading-relaxed", className)}
    {...props}
  >
    {children}
  </p>
))
AlertDescription.displayName = "AlertDescription"

export { Alert, AlertTitle, AlertDescription }

