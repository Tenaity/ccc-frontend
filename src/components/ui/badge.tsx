import * as React from "react"
import { cva, type VariantProps } from "class-variance-authority"

import { cn } from "@/lib/utils"

const badgeVariants = cva(
  "inline-flex items-center gap-1 rounded-full border border-transparent px-2.5 py-0.5 text-xs font-semibold transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2",
  {
    variants: {
      variant: {
        default: "bg-primary text-primary-foreground shadow-sm",
        secondary: "bg-secondary text-secondary-foreground",
        destructive: "bg-destructive text-destructive-foreground shadow-sm",
        outline: "border-border/70 bg-background text-foreground shadow-sm",
        muted: "bg-muted text-muted-foreground",
        leader:
          "border-emerald-200 bg-emerald-50 text-emerald-900 shadow-[inset_0_1px_0_rgba(255,255,255,0.6)]",
        night:
          "border-slate-800 bg-slate-900 text-slate-50 shadow-[0_1px_0_rgba(15,23,42,0.55)]",
        pgd: "border-rose-200 bg-rose-50 text-rose-900 shadow-[inset_0_1px_0_rgba(255,255,255,0.65)]",
        duplicate:
          "border-destructive/70 text-destructive ring-2 ring-destructive/60 shadow-[0_0_0_1px_rgba(248,113,113,0.35)]",
        fixed: "border-primary/40 bg-primary/10 text-primary",
        off: "border-border/70 bg-muted text-muted-foreground",
      },
    },
    defaultVariants: {
      variant: "default",
    },
  }
)

export interface BadgeProps
  extends React.HTMLAttributes<HTMLDivElement>,
    VariantProps<typeof badgeVariants> {}

const Badge = React.forwardRef<HTMLDivElement, BadgeProps>(
  ({ className, variant, ...props }, ref) => {
    return (
      <div
        ref={ref}
        className={cn(badgeVariants({ variant }), className)}
        {...props}
      />
    )
  }
)
Badge.displayName = "Badge"

export { Badge, badgeVariants }
