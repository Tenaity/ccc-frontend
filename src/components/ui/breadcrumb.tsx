import * as React from "react"

import { cn } from "@/lib/utils"

const Breadcrumb = React.forwardRef<
  HTMLElement,
  React.ComponentPropsWithoutRef<"nav">
>(({ className, ...props }, ref) => (
  <nav
    ref={ref}
    aria-label="Breadcrumb"
    className={cn("w-full", className)}
    {...props}
  />
))
Breadcrumb.displayName = "Breadcrumb"

const BreadcrumbList = React.forwardRef<
  HTMLOListElement,
  React.ComponentPropsWithoutRef<"ol">
>(({ className, ...props }, ref) => (
  <ol
    ref={ref}
    className={cn(
      "flex flex-wrap items-center gap-1.5 text-sm text-muted-foreground",
      className
    )}
    {...props}
  />
))
BreadcrumbList.displayName = "BreadcrumbList"

const BreadcrumbItem = React.forwardRef<
  HTMLLIElement,
  React.ComponentPropsWithoutRef<"li">
>(({ className, ...props }, ref) => (
  <li
    ref={ref}
    className={cn("inline-flex items-center gap-1", className)}
    {...props}
  />
))
BreadcrumbItem.displayName = "BreadcrumbItem"

const BreadcrumbLink = React.forwardRef<
  HTMLAnchorElement,
  React.ComponentPropsWithoutRef<"a">
>(({ className, ...props }, ref) => (
  <a
    ref={ref}
    className={cn(
      "inline-flex items-center gap-1 transition-colors hover:text-foreground",
      className
    )}
    {...props}
  />
))
BreadcrumbLink.displayName = "BreadcrumbLink"

const BreadcrumbPage = React.forwardRef<
  HTMLSpanElement,
  React.ComponentPropsWithoutRef<"span">
>(({ className, ...props }, ref) => (
  <span
    ref={ref}
    aria-current="page"
    className={cn("font-medium text-foreground", className)}
    {...props}
  />
))
BreadcrumbPage.displayName = "BreadcrumbPage"

const BreadcrumbSeparator = ({
  className,
  children,
  ...props
}: React.ComponentPropsWithoutRef<"span">) => (
  <span
    role="presentation"
    className={cn("text-muted-foreground", className)}
    {...props}
  >
    {children ?? "/"}
  </span>
)
BreadcrumbSeparator.displayName = "BreadcrumbSeparator"

const BreadcrumbEllipsis = ({
  className,
  ...props
}: React.ComponentPropsWithoutRef<"span">) => (
  <span
    className={cn("inline-flex items-center", className)}
    {...props}
  >
    <span aria-hidden>â€¦</span>
    <span className="sr-only">More</span>
  </span>
)
BreadcrumbEllipsis.displayName = "BreadcrumbEllipsis"

export {
  Breadcrumb,
  BreadcrumbList,
  BreadcrumbItem,
  BreadcrumbLink,
  BreadcrumbPage,
  BreadcrumbSeparator,
  BreadcrumbEllipsis,
}

