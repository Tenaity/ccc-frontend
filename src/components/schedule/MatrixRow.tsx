// src/components/schedule/MatrixRow.tsx
import React from "react";
import type { Staff, Assignment } from "@/types";
import { fmtYMD, getDow, isWeekend } from "@/utils/date";
import Badge from "@/components/Badge";
import {
  isNightLeader,
  isDayLeader,
  isNightTD,
  isNightPGD,
} from "@/utils/schedule";
import { cn } from "@/lib/utils";
import { TableCell, TableRow } from "@/components/ui/table";
import { Tooltip, TooltipContent, TooltipTrigger } from "@/components/ui/tooltip";
import { ShieldCheck } from "lucide-react";

// Parse tiện lợi từ notes: "[CODE:1978][RANK:1]"
function parseMeta(notes?: string | null) {
    if (!notes) return { code: null as string | null, rank: null as 1 | 2 | null };
    const code = notes.match(/\[CODE:(\d+)\]/)?.[1] ?? null;
    const rankStr = notes.match(/\[RANK:(1|2)\]/)?.[1] ?? null;
    const rank = (rankStr === "1" || rankStr === "2") ? (Number(rankStr) as 1 | 2) : null;
    return { code, rank };
}

export default function MatrixRow({
  staff,
  index,
  year,
  month,
  days,
  assignmentIndex,
  summariesByStaffId,
  fixedByDayStaff,
  offByDayStaff,
  leaderCountsByDay,
  onEditCell,
}: {
  staff: Staff;
  index: number;
  year: number;
  month: number;
  days: number[];
  assignmentIndex: Map<
    string,
    { code: Assignment["shift_code"]; position: Assignment["position"] | null }
  >;
  summariesByStaffId: Map<
    number,
    {
      counts: Record<string, number>;
      credit: number;
      dayCount: number;
      nightCount: number;
    }
  >;
  fixedByDayStaff: Map<string, boolean>;
  offByDayStaff: Map<string, boolean>;
  leaderCountsByDay: Map<number, { day: number; night: number }>;
  onEditCell?: (staff: Staff, day: number) => void;
}) {
  const sum =
    summariesByStaffId.get(staff.id) ||
    { counts: {}, credit: 0, dayCount: 0, nightCount: 0 };
  const meta = parseMeta(staff.notes);
  const displayId = meta.code ?? String(staff.id);

  const stripeClass = index % 2 === 0 ? "bg-primary/5" : "bg-primary/10";

  const RankChip = meta.rank ? (
    <span
      className={cn(
        "ml-2 inline-flex items-center rounded-full border px-1.5 text-[11px] font-semibold",
        meta.rank === 1
          ? "border-primary/70 bg-primary/10 text-primary"
          : "border-dashed border-primary/50 bg-primary/5 text-primary"
      )}
    >
      Rank {meta.rank}
    </span>
  ) : null;

  return (
    <TableRow
      className={cn(
        "hover:bg-primary/15",
        stripeClass,
      )}
    >
      <th
        scope="row"
        className="sticky left-0 z-20 min-w-[220px] border-r border-border/60 bg-background px-4 py-3 text-left shadow-[4px_0_12px_-8px_rgba(15,23,42,0.25)] backdrop-blur-sm"
      >
        <div className="flex items-baseline justify-between gap-3">
          <div>
            <div className="text-sm font-semibold text-foreground">
              {staff.full_name}
            </div>
            <div className="flex items-center gap-2 text-xs text-muted-foreground">
              <span>{staff.role}</span>
              <span className="rounded-md border border-primary/20 bg-primary/10 px-1.5 py-0.5 font-medium text-primary">
                #{displayId}
              </span>
              {RankChip}
            </div>
          </div>
          {meta.rank === 1 ? (
            <ShieldCheck className="h-4 w-4 text-primary" aria-hidden="true" />
          ) : null}
        </div>
      </th>

      {days.map((day) => {
        const dateKey = fmtYMD(year, month, day);
        const cell = assignmentIndex.get(`${staff.id}|${dateKey}`);
        const code = (cell?.code ?? "") as Assignment["shift_code"] | "";
        const position =
          (cell?.position ?? null) as Assignment["position"] | null;
        const weekend = isWeekend(getDow(year, month, day));

        const marker = { shift_code: code, position, role: staff.role };
        const leaderDay = isDayLeader(marker);
        const leaderNight = isNightLeader(marker);
        const tdNight = isNightTD(marker);
        const pgdNight = isNightPGD(marker);

        const leaderBucket = leaderCountsByDay.get(day);
        const duplicateDayLeader = leaderDay && (leaderBucket?.day ?? 0) > 1;
        const duplicateNightLeader = leaderNight && (leaderBucket?.night ?? 0) > 1;

        let badgeVariant: React.ComponentProps<typeof Badge>["variant"] = "default";
        if (leaderDay) badgeVariant = duplicateDayLeader ? "duplicate" : "leader";
        else if (leaderNight)
          badgeVariant = duplicateNightLeader ? "duplicate" : "night";
        else if (tdNight) badgeVariant = "night";
        else if (pgdNight) badgeVariant = "pgd";
        else if (position === "PGD") badgeVariant = "pgd";

        const key = `${staff.id}|${dateKey}`;
        const isFixed = fixedByDayStaff.has(key);
        const isOff = offByDayStaff.has(key);

        const tooltipSource = isOff
          ? "Nguồn: Off"
          : isFixed
            ? "Nguồn: Fixed"
            : "Nguồn: Engine";

        const displayCode = isOff ? "OFF" : code || "Trống";

        return (
          <TableCell
            key={day}
            className={cn(
              "relative h-16 min-w-[64px] border-r border-border/40 px-1 py-1 text-center",
              weekend && "bg-amber-50/60",
              !isFixed && !isOff && onEditCell && "cursor-pointer hover:bg-primary/5"
            )}
            onClick={() => {
              if (!isFixed && !isOff && onEditCell) {
                onEditCell(staff, day);
              }
            }}
          >
            <Tooltip>
              <TooltipTrigger asChild>
                <div className="flex h-full w-full items-center justify-center">
                  {isOff ? (
                    <Badge code="OFF" variant="off" />
                  ) : (
                    <Badge
                      code={code || ""}
                      crown={leaderDay || leaderNight}
                      variant={isFixed ? "fixed" : badgeVariant}
                      pinned={isFixed}
                      rank={meta.rank || undefined}
                    />
                  )}
                </div>
              </TooltipTrigger>
              <TooltipContent className="max-w-xs text-left">
                <p className="font-semibold">{staff.full_name}</p>
                <p className="text-xs text-muted-foreground">
                  {dateKey} · {displayCode}
                </p>
                {position ? (
                  <p className="text-xs">Vị trí: {position}</p>
                ) : null}
                <p className="text-xs">{tooltipSource}</p>
                <p className="text-xs">Credit tháng: {sum.credit}</p>
              </TooltipContent>
            </Tooltip>
          </TableCell>
        );
      })}

      <TableCell className="text-center text-sm font-medium text-foreground">
        {sum.counts["CA1"] || 0}
      </TableCell>
      <TableCell className="text-center text-sm font-medium text-foreground">
        {sum.counts["CA2"] || 0}
      </TableCell>
      <TableCell className="text-center text-sm font-medium text-foreground">
        {sum.counts["K"] || 0}
      </TableCell>
      <TableCell className="text-center text-sm font-medium text-foreground">
        {sum.counts["HC"] || 0}
      </TableCell>
      <TableCell className="text-center text-sm font-medium text-foreground">
        {sum.counts["Đ"] || 0}
      </TableCell>
      <TableCell className="text-center text-sm font-medium text-foreground">
        {sum.counts["P"] || 0}
      </TableCell>
      <TableCell className="bg-muted/40 text-center text-sm font-semibold text-foreground">
        {sum.dayCount || 0}
      </TableCell>
      <TableCell className="bg-muted/40 text-center text-sm font-semibold text-foreground">
        {sum.nightCount || 0}
      </TableCell>
      <TableCell className="bg-muted/60 text-center text-sm font-bold text-foreground">
        {sum.credit || 0}
      </TableCell>
    </TableRow>
  );
}