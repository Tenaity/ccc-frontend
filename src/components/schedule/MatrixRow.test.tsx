import { render } from "@testing-library/react"
import { expect, test } from "vitest"

import MatrixRow from "@/components/schedule/MatrixRow"
import { TooltipProvider } from "@/components/ui/tooltip"
import type { Assignment, Staff } from "@/types"

const year = 2025
const month = 5
const days = [1, 2, 3]

const staff: Staff = {
  id: 42,
  full_name: "Nguyễn Văn A",
  role: "TC",
  can_night: true,
  base_quota: 18,
  notes: "[CODE:1978][RANK:1]",
}

const dayKey = (day: number) =>
  `${year}-${String(month).padStart(2, "0")}-${String(day).padStart(2, "0")}`

const assignmentIndex = new Map<
  string,
  { code: Assignment["shift_code"]; position: Assignment["position"] | null }
>()
assignmentIndex.set(`${staff.id}|${dayKey(1)}`, { code: "K", position: "TD" })
assignmentIndex.set(`${staff.id}|${dayKey(2)}`, { code: "Đ", position: "TD" })

const summariesByStaffId = new Map<number, {
  counts: Record<string, number>
  credit: number
  dayCount: number
  nightCount: number
}>()
summariesByStaffId.set(staff.id, {
  counts: { CA1: 1, CA2: 0, K: 1, HC: 0, Đ: 1, P: 1 },
  credit: 12,
  dayCount: 2,
  nightCount: 1,
})

const fixedByDayStaff = new Map<string, boolean>([
  [`${staff.id}|${dayKey(1)}`, true],
])

const offByDayStaff = new Map<string, boolean>([
  [`${staff.id}|${dayKey(3)}`, true],
])

const leaderCountsByDay = new Map<number, { day: number; night: number }>([
  [1, { day: 2, night: 0 }],
  [2, { day: 0, night: 1 }],
  [3, { day: 0, night: 0 }],
])

test("renders schedule matrix row with badges and totals", () => {
  const { asFragment } = render(
    <TooltipProvider>
      <table>
        <tbody>
          <MatrixRow
            staff={staff}
            index={0}
            year={year}
            month={month}
            days={days}
            assignmentIndex={assignmentIndex}
            summariesByStaffId={summariesByStaffId}
            fixedByDayStaff={fixedByDayStaff}
            offByDayStaff={offByDayStaff}
            leaderCountsByDay={leaderCountsByDay}
          />
        </tbody>
      </table>
    </TooltipProvider>,
  )

  expect(asFragment()).toMatchSnapshot()
})
