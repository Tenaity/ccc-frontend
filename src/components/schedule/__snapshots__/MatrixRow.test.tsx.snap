// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MatrixRow > matches snapshot with duplicate leaders and rank badge 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="border-b border-border/60 transition-colors hover:bg-primary/15 bg-primary/5"
      >
        <th
          class="sticky left-0 z-20 min-w-[220px] border-r border-border/60 bg-background/95 px-4 py-3 text-left shadow-[4px_0_12px_-8px_rgba(15,23,42,0.25)] backdrop-blur-sm"
          scope="row"
        >
          <div
            class="flex items-baseline justify-between gap-3"
          >
            <div>
              <div
                class="text-sm font-semibold text-foreground"
              >
                Nguyen Van A
              </div>
              <div
                class="flex items-center gap-2 text-xs text-muted-foreground"
              >
                <span>
                  TC
                </span>
                <span
                  class="rounded-md border border-primary/20 bg-primary/10 px-1.5 py-0.5 font-medium text-primary"
                >
                  #1978
                </span>
                <span
                  class="ml-2 inline-flex items-center rounded-full border px-1.5 text-[11px] font-semibold border-primary/70 bg-primary/10 text-primary"
                >
                  Rank 1
                </span>
              </div>
            </div>
            <svg
              aria-hidden="true"
              class="lucide lucide-shield-check h-4 w-4 text-primary"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 13c0 5-3.5 7.5-7.66 8.95a1 1 0 0 1-.67-.01C7.5 20.5 4 18 4 13V6a1 1 0 0 1 1-1c2 0 4.5-1.2 6.24-2.72a1.17 1.17 0 0 1 1.52 0C14.51 3.81 17 5 19 5a1 1 0 0 1 1 1z"
              />
              <path
                d="m9 12 2 2 4-4"
              />
            </svg>
          </div>
        </th>
        <td
          class="p-3 align-middle text-sm relative h-16 min-w-[64px] border-r border-border/40 px-1 py-1 text-center bg-amber-50/60 cursor-pointer hover:bg-primary/5"
        >
          <div
            class="flex h-full w-full items-center justify-center"
            data-state="closed"
          >
            <div
              class="focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 inline-flex min-w-[3rem] items-center justify-center gap-1 rounded-lg px-2 py-1 text-xs font-semibold uppercase tracking-tight transition-colors bg-emerald-100 ring-2 ring-destructive/70 text-destructive shadow-[0_0_0_1px_rgba(248,113,113,0.3)] border border-transparent"
            >
              <span>
                K
              </span>
              <svg
                aria-hidden="true"
                class="lucide lucide-crown h-3.5 w-3.5 text-amber-500"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.562 3.266a.5.5 0 0 1 .876 0L15.39 8.87a1 1 0 0 0 1.516.294L21.183 5.5a.5.5 0 0 1 .798.519l-2.834 10.246a1 1 0 0 1-.956.734H5.81a1 1 0 0 1-.957-.734L2.02 6.02a.5.5 0 0 1 .798-.519l4.276 3.664a1 1 0 0 0 1.516-.294z"
                />
                <path
                  d="M5 21h14"
                />
              </svg>
            </div>
          </div>
        </td>
        <td
          class="p-3 align-middle text-sm relative h-16 min-w-[64px] border-r border-border/40 px-1 py-1 text-center bg-amber-50/60 cursor-pointer hover:bg-primary/5"
        >
          <div
            class="flex h-full w-full items-center justify-center"
            data-state="closed"
          >
            <div
              class="focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 inline-flex min-w-[3rem] items-center justify-center gap-1 rounded-lg px-2 py-1 text-xs font-semibold uppercase tracking-tight transition-colors bg-rose-100 ring-2 ring-destructive/70 text-destructive shadow-[0_0_0_1px_rgba(248,113,113,0.3)] border border-transparent"
            >
              <span>
                Đ
              </span>
              <svg
                aria-hidden="true"
                class="lucide lucide-crown h-3.5 w-3.5 text-amber-500"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.562 3.266a.5.5 0 0 1 .876 0L15.39 8.87a1 1 0 0 0 1.516.294L21.183 5.5a.5.5 0 0 1 .798.519l-2.834 10.246a1 1 0 0 1-.956.734H5.81a1 1 0 0 1-.957-.734L2.02 6.02a.5.5 0 0 1 .798-.519l4.276 3.664a1 1 0 0 0 1.516-.294z"
                />
                <path
                  d="M5 21h14"
                />
              </svg>
            </div>
          </div>
        </td>
        <td
          class="p-3 align-middle text-sm relative h-16 min-w-[64px] border-r border-border/40 px-1 py-1 text-center"
        >
          <div
            class="flex h-full w-full items-center justify-center"
            data-state="closed"
          >
            <div
              class="focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 inline-flex min-w-[3rem] items-center justify-center gap-1 rounded-lg px-2 py-1 text-xs font-semibold uppercase tracking-tight transition-colors bg-blue-100 text-blue-900 border border-primary/60"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-pin h-3.5 w-3.5 text-primary"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 17v5"
                />
                <path
                  d="M9 10.76a2 2 0 0 1-1.11 1.79l-1.78.9A2 2 0 0 0 5 15.24V16a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1v-.76a2 2 0 0 0-1.11-1.79l-1.78-.9A2 2 0 0 1 15 10.76V7a1 1 0 0 1 1-1 2 2 0 0 0 0-4H8a2 2 0 0 0 0 4 1 1 0 0 1 1 1z"
                />
              </svg>
              <span>
                CA1
              </span>
            </div>
          </div>
        </td>
        <td
          class="p-3 align-middle text-center text-sm font-medium text-foreground"
        >
          2
        </td>
        <td
          class="p-3 align-middle text-center text-sm font-medium text-foreground"
        >
          1
        </td>
        <td
          class="p-3 align-middle text-center text-sm font-medium text-foreground"
        >
          1
        </td>
        <td
          class="p-3 align-middle text-center text-sm font-medium text-foreground"
        >
          0
        </td>
        <td
          class="p-3 align-middle text-center text-sm font-medium text-foreground"
        >
          1
        </td>
        <td
          class="p-3 align-middle text-center text-sm font-medium text-foreground"
        >
          0
        </td>
        <td
          class="p-3 align-middle bg-muted/40 text-center text-sm font-semibold text-foreground"
        >
          4
        </td>
        <td
          class="p-3 align-middle bg-muted/40 text-center text-sm font-semibold text-foreground"
        >
          1
        </td>
        <td
          class="p-3 align-middle bg-muted/60 text-center text-sm font-bold text-foreground"
        >
          7
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
